src/VisualStudio/Core/Def/Implementation/PullMemberUp/MainDialog/BaseTypeTreeNodeViewModel.cs

```diff
@@ -0,0 +1,63 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.VisualStudio.Language.Intellisense;
+using static Microsoft.VisualStudio.LanguageServices.Implementation.ExtractInterface.ExtractInterfaceDialogViewModel;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.PullMemberUp.MainDialog
+{
+    /// <summary>
+    /// View model used to represent and display the inheritance graph as a tree. This tree is constructed by breadth first searching.
+    /// If one type is the common base type of several other types, it will be showed multiple time.
+    /// </summary>
+    internal class BaseTypeTreeNodeViewModel : MemberSymbolViewModel 
+    {
+        /// <summary>
+        /// Base types of this tree node
+        /// </summary>
+        public ImmutableArray<BaseTypeTreeNodeViewModel> BaseTypeNodes { get; private set; }
+
+        public bool IsExpanded { get; set; }
+
+        /// <summary>
+        /// Content of the tooltip.
+        /// </summary>
+        public string Namespace => string.Format(ServicesVSResources.Namespace_0,
+            MemberSymbol.ContainingNamespace?.ToDisplayString() ?? ServicesVSResources.Namespace_global);
+
+        private BaseTypeTreeNodeViewModel(INamedTypeSymbol node, IGlyphService glyphService) : base(node, glyphService)
+        {
+        }
+
+        /// <summary>
+        /// Use breadth first search to create the inheritance tree. Only types in the solution will be included in the tree.
+        /// </summary>
+        internal static BaseTypeTreeNodeViewModel CreateBaseTypeTree(INamedTypeSymbol root, IGlyphService glyphService)
+        {
+            var rootTreeNode = new BaseTypeTreeNodeViewModel(root, glyphService) { IsChecked = false, IsExpanded = true};
+            var queue = new Queue<BaseTypeTreeNodeViewModel>();
+            queue.Enqueue(rootTreeNode);
+            while (queue.Any())
+            {
+                var currentTreeNode = queue.Dequeue();
+                var currentTypeSymbol = currentTreeNode.MemberSymbol as INamedTypeSymbol;
+
+                currentTreeNode.BaseTypeNodes = currentTypeSymbol.Interfaces.Concat(currentTypeSymbol.BaseType).
```

Comment:
@CyrusNajmabadi  I was trying to apply the function in MemberAndDestinationValidator. But it seems like I need a cancellation token to check if the symbol is generated code. Is there anyway to do it here?

Reply:
Yes.  Instead of synchronously computing this, create your own cancellation token kick off the work to compute this.  Cancel the token if the dialog is closed.

Reply:
While this is being computed, show a spinner in the 'destination' location.

b

Reply:
Note: the whole way the dialog works seems odd.  When the dialog is created, it should make one cancellation token source, and that will be the cancellation token for all async work in the dialog.  The token is always canceled when the dialog is dismissed.  

--

Note: i would work with @jasonmalinowski who can help you on how to properly create a dialog that does async workin a clear and simple manner.

Reply:
Have an appointment with Jason tomorrow morning, we will discuss it

Reply:
We chatted and here's what we ended up with:

The use of AsyncLazy here doesn't really make sense, because the "lazy" bit is being ignored: we're creating AsyncLazies, and then starting Tasks that are immediately pulling on the value. So it's easier just to start regular Tasks directly, all associated with the dialog's overall CancellationTokenSource, and those will cancel off when the dialog is dismissed.

The other "benefit" of AsyncLazy would be the merging of cancellation tokens, but we're not using that either. You could imagine that clicking OK would take the AsyncLazies that were needed and continue to use them after the dialog is closed; in that case it'd be nice because while the other thing cancelling off the work that wasn't used, it won't implicitly cancel off the work that was used. But we're not doing that here -- it's clear that the dialog going away *should* mean all the work is done; if it's not something is up.

I also suggested to avoid the use of await-style async when you click the "select dependents" button, and instead just grab the Task we started and just .Wait() it within a IWaitIndicator. Right now as the user if you click the button and it has to think, it'll take its time and...eventually....maybe, do something. If you misclicked and didn't want that, your only recourse is to either wait until it does it's thing and stop, or pray you can click OK before it comes back. If we just do a IWaitIndicator, we'll tell the user we're thinking (with an actual message instead of a nondescript spinner), and they'll have a cancel button they can click. In this case, it would mean we're simply cancelling waiting on the result of the Tasks we started when the dialog opened. The underlying computations keep going, but you're not blocked waiting for them if you decide you didn't want that.